In [18]:
import pandas as pd
import numpy as np
import seaborn as sns

import joblib
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_digits

In [19]:
# Load the uploaded dataset
file_path = "/Users/adamkhay/Desktop/data engineering/cleaned_data.csv"
df = pd.read_csv(file_path)

# Display basic info
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   StudentID                   10000 non-null  int64  
 1   CGPA                        10000 non-null  float64
 2   Internships                 10000 non-null  int64  
 3   Projects                    10000 non-null  int64  
 4   Workshops_Certifications    10000 non-null  int64  
 5   AptitudeTestScore           10000 non-null  int64  
 6   SoftSkillsRating            10000 non-null  float64
 7   SSC_Marks                   10000 non-null  int64  
 8   HSC_Marks                   10000 non-null  int64  
 9   PlacementStatus             10000 non-null  object 
 10  has_done_extracurricular    10000 non-null  int64  
 11  has_done_placementtraining  10000 non-null  int64  
dtypes: float64(2), int64(9), object(1)
memory usage: 937.6+ KB


In [20]:
# Display first few rows
df.head()

StudentID  CGPA  Internships  Projects  Workshops_Certifications  \
0          1   7.5            1         1                         1   
1          2   8.9            0         3                         2   
2          3   7.3            1         2                         2   
3          4   7.5            1         1                         2   
4          5   8.3            1         2                         2   

   AptitudeTestScore  SoftSkillsRating  SSC_Marks  HSC_Marks PlacementStatus  \
0                 65               4.4         61         79       NotPlaced   
1                 90               4.0         78         82          Placed   
2                 82               4.8         79         80       NotPlaced   
3                 85               4.4         81         80          Placed   
4                 86               4.5         74         88          Placed   

   has_done_extracurricular  has_done_placementtraining  
0                         0                           0  
1                         1                           1  
2                         1                           0  
3                         1                           1  
4                         1                           1

In [21]:
feature_cols = ["has_done_extracurricular", "has_done_placementtraining", "SoftSkillsRating"]
prediction_col = "PlacementStatus"


In [22]:
#set the prediction column and the feature columns 

X = df[feature_cols]
y = df[prediction_col]

In [23]:
# Train-Validation-Test Split - First split
X_train, X_temp, y_train, y_temp = train_test_split(X, y, random_state=42, test_size=0.3, stratify=y)

In [24]:
X_temp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3000 entries, 4258 to 7089
Data columns (total 3 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   has_done_extracurricular    3000 non-null   int64  
 1   has_done_placementtraining  3000 non-null   int64  
 2   SoftSkillsRating            3000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 93.8 KB


In [25]:
# Train-Validation-Test Split - First split
X_train, X_temp, y_train, y_temp = train_test_split(X, y, random_state=42, test_size=0.3, stratify=y)

In [26]:
X_temp.head()

has_done_extracurricular  has_done_placementtraining  SoftSkillsRating
4258                         0                           0               3.9
1687                         1                           0               4.8
6130                         1                           1               4.8
5761                         0                           1               4.1
9587                         1                           0               4.6

In [27]:
y_temp.info()

<class 'pandas.core.series.Series'>
Index: 3000 entries, 4258 to 7089
Series name: PlacementStatus
Non-Null Count  Dtype 
--------------  ----- 
3000 non-null   object
dtypes: object(1)
memory usage: 46.9+ KB


In [28]:
y_temp.head()

4258    NotPlaced
1687    NotPlaced
6130    NotPlaced
5761    NotPlaced
9587       Placed
Name: PlacementStatus, dtype: object

In [29]:
# Train-Validation-Test Split - Second split
X_test, X_val, y_test, y_val = train_test_split(X_temp, y_temp, random_state=42, test_size=0.5, stratify=y_temp)

In [30]:
print(X_test.shape)
X_test.head()

(1500, 3)


has_done_extracurricular  has_done_placementtraining  SoftSkillsRating
3891                         0                           1               3.8
7136                         1                           1               4.4
1823                         0                           0               4.0
1560                         1                           1               3.9
104                          0                           0               4.1

In [31]:
print(X_val.shape)
X_val.head()

(1500, 3)


has_done_extracurricular  has_done_placementtraining  SoftSkillsRating
9232                         0                           1               3.8
4178                         1                           1               4.8
6582                         1                           1               4.6
4405                         1                           1               4.8
388                          0                           0               4.4

In [32]:
print(y_test.shape)
y_test.head()

(1500,)


3891    NotPlaced
7136    NotPlaced
1823    NotPlaced
1560    NotPlaced
104     NotPlaced
Name: PlacementStatus, dtype: object

In [33]:
print(y_val.shape)
y_val.head()

(1500,)


9232    NotPlaced
4178       Placed
6582       Placed
4405    NotPlaced
388        Placed
Name: PlacementStatus, dtype: object

In [34]:
# Load the saved model
loaded_model = joblib.load("digits_model.joblib")

In [35]:
# Make predictions on the validation set
y_val_pred = loaded_model.predict(X_val)

In [36]:
y_val_pred

array(['NotPlaced', 'Placed', 'Placed', ..., 'NotPlaced', 'Placed',
       'NotPlaced'], dtype=object)

In [37]:
from sklearn.metrics import accuracy_score

In [38]:
accuracy_score(y_val,y_val_pred)

0.7813333333333333

Storing the prediction from the validation set in database. in a new schema named ‘analytics’.

In [42]:
!pip install psycopg2-binary

In [43]:
import psycopg2
print("psycopg2 installed successfully!")

psycopg2 installed successfully!


In [44]:
#!pip install sqlalchemy

In [45]:
from sqlalchemy import create_engine
print("SQLAlchemy imported successfully!")

SQLAlchemy imported successfully!


In [46]:
host = r'127.0.0.1' # denotes that the db in a local installation
db = r'MSDS610' # db we just created
user = r'postgres' # using the postgres user for this demo
pw = r'adamkhay' # this is the password established during installation
port = r'5432' # default port estabalished during install
schema = r'raw' # schema we just created

In [47]:
db_conn = create_engine("postgresql://{}:{}@{}:{}/{}".format(user, pw, host, port, db))

In [48]:
sql="select tables.table_name from information_schema.tables where (table_schema ='"+schema+"')order by 1;"
tbl_df = pd.read_sql(sql, db_conn, index_col=None)
tbl_df

table_name
0  placement_data

In [49]:
table_name = r'placement_data'

In [50]:
df.to_sql(table_name, con=db_conn, if_exists='replace', index=False, schema=schema, chunksize=1000, method='multi')

10000

In [51]:
tbl_df = pd.read_sql(sql, db_conn, index_col=None)
tbl_df

table_name
0  placement_data

In [52]:
sql=r'SELECT * FROM ' + schema + '.' + table_name
df_check = pd.read_sql(sql, db_conn, index_col=None)

In [55]:
table_name = r'placementdata_cleaned'
schema = r'analytics' # schema 

df.to_sql(table_name, con=db_conn, if_exists='replace', index=False, schema=schema, chunksize=1000, method='multi')

10000

Recap of the Prediction from the Validation Set:

We notice that the Test Accuracy is 78%, similar to the Validation Accuracy that is 78% as well.
The model's ability to generalize to new data is demonstrated by the fact that both scores are equal.
This suggests that there is no overfitting because the model functions consistently across various unknown data, demonstrating that it successfully learns patterns rather than memorizing the training data.
Stable Generalization: The equal accuracy of the test and validation sets, which were divided proportionately but at random (stratify=y), indicates that the dataset is well-balanced.
Real-World Readiness: If the model's distribution matches your dataset, it should function similarly when applied to brand-new student placement data.
